File: src/main/java/com/now/nowbot/model/beatmapParse/HitObject.java
package com.now.nowbot.model.beatmapParse;

import com.now.nowbot.model.beatmapParse.hitObject.HitObjectPosition;
import com.now.nowbot.model.beatmapParse.hitObject.HitObjectSound;
import com.now.nowbot.model.beatmapParse.hitObject.HitObjectType;
import com.now.nowbot.model.beatmapParse.hitObject.SliderAttr;

public class HitObject {
    HitObjectPosition position;
    HitObjectType  type;
    HitObjectSound sound;
    Integer        startTime;
    Integer        endTime;

    Integer column;

    /**
     * 仅 type == Slider 时存在
     */
    SliderAttr sliderAttr;

    public HitObject() {
    }

    public HitObject(int x, int y, HitObjectType type, int time, int end) {
        position = new HitObjectPosition(x, y);
        startTime = time;
        endTime = end;
    }

    public HitObjectPosition getPosition() {
        return position;
    }

    public void setPosition(HitObjectPosition position) {
        this.position = position;
    }

    public HitObjectType getType() {
        return type;
    }

    public void setType(HitObjectType type) {
        this.type = type;
    }

    public void setSound(HitObjectSound sound) {
        this.sound = sound;
    }

    public HitObjectSound getSound() {
        return sound;
    }

    public int getStartTime() {
        return startTime;
    }

    public void setStartTime(int time) {
        this.startTime = time;
    }

    public int getEndTime() {
        return endTime;
    }

    public void setEndTime(int time) {
        this.endTime = time;
    }

    public int getColumn() {
        return column;
    }

    public void setColumn(int column) {
        this.column = column;
    }
}

File: src/main/java/com/now/nowbot/model/beatmapParse/HitObject.kt
package com.now.nowbot.model.beatmapParse

import com.now.nowbot.model.beatmapParse.hitObject.HitObjectPosition
import com.now.nowbot.model.beatmapParse.hitObject.HitObjectSound
import com.now.nowbot.model.beatmapParse.hitObject.HitObjectType
import com.now.nowbot.model.beatmapParse.hitObject.SliderAttr

class HitObject {
    var position: HitObjectPosition = HitObjectPosition()
    var type: HitObjectType = HitObjectType.DEFAULT
    var sound: HitObjectSound = HitObjectSound.DEFAULT
    var startTime: Int = 0

    var endTime: Int = 0
        get() = if (field > 0) field else startTime

    var column: Int = 0

    /**
     * 仅 type == Slider 时存在
     */
    var sliderAttr: SliderAttr? = null

    constructor()

    constructor(x: Int, y: Int, type: HitObjectType, time: Int, end: Int) {
        this.position = HitObjectPosition(x, y)
        this.startTime = time
        this.endTime = end
        this.type = type
    }
}
